plutus-wallet-api/ledger/Ledger/Typed/Tx.hs

```diff
@@ -0,0 +1,419 @@
+-- type errors at validation time.
+module Ledger.Typed.Tx where
+
+import qualified Ledger.Ada                   as Ada
+import           Ledger.Crypto
+import qualified Ledger.Interval              as Interval
+import           Ledger.Scripts
+import           Ledger.Slot
+import           Ledger.Tx                    hiding (scriptAddress)
+import qualified Ledger.Tx                    as Tx
+import qualified Ledger.Validation            as Validation
+import qualified Ledger.Value                 as Value
+
+import           Ledger.Typed.TypeUtils
+
+import qualified Language.PlutusCore          as PLC
+import qualified Language.PlutusCore.Pretty   as PLC
+
+import           Language.PlutusTx
+import qualified Language.PlutusTx.Builtins   as PlutusTx
+import           Language.PlutusTx.Lift       as Lift
+import           Language.PlutusTx.Lift.Class as Lift
+
+import qualified Language.PlutusIR.Compiler   as PIR
+
+import           Data.Coerce
+import           Data.Kind
+import           Data.List                    (foldl')
+import           Data.Proxy
+import qualified Data.Set                     as Set
+
+import           Control.Monad.Except
+
+import qualified Data.Singletons              as Sing
+import qualified Data.Singletons.Prelude.List as Sing
+
+-- | A class that associates a type standing for a connection type with two types, the type of the redeemer
+-- and the data script for that connection type.
+class ScriptType (a :: Type) where
+    -- | The type of the redeemers of this connection type.
+    type RedeemerType a :: Type
+    -- | The type of the data of this connection type.
+    type DataType a :: Type
+
+    -- Defaults
+    type instance RedeemerType a = ()
+    type instance DataType  a = ()
+
+-- | The type of validators for the given connection type.
+type ValidatorType (a :: Type) = DataType a -> RedeemerType a -> Validation.PendingTx -> Bool
+
+-- | The type of a connection.
+data ScriptInstance (a :: Type) where
+    Validator :: ScriptType a => CompiledCode (ValidatorType a) -> ScriptInstance a
+
+-- | Get the address for a script instance.
+scriptAddress :: ScriptInstance a -> Address
+scriptAddress = Tx.scriptAddress . validatorScript
+
+-- | Get the validator script for a script instance.
+validatorScript :: ScriptInstance a -> ValidatorScript
+validatorScript (Validator vc) = ValidatorScript $ fromCompiledCode vc
+
+-- | Defunctionalized symbol for use with 'Apply'.
+data SealedDataTypeSym :: Type Sing.~> Type
+type instance Sing.Apply SealedDataTypeSym a = PlutusTx.Sealed (HashedDataScript (DataType a))
+-- | Defunctionalized symbol for use with 'Apply'.
+data DataTypeSym :: Type Sing.~> Type
+type instance Sing.Apply DataTypeSym a = DataType a
+
+-- | The type of a redeemer function for the given output types and input type.
+type RedeemerFunctionType (outs :: [Type]) (inn :: Type) = Uncurry (Sing.Map SealedDataTypeSym outs) (RedeemerType inn)
```

Comment:
I'm not sure I understand? I tried just having a simple `Map` family, but you can't use it with `SealedDataType` because it's a type synonym. AFAIK the way to do a mapping with a type synonym is defuncitonalization.

Reply:
I mean, you could inline the `Map SealedDataTypeSym` part and provide the specialized `MapSealedDataTypeSym` type family. Defunctionalization is useful when you're applying different functions, but in this case we have only one function and do not need too much to deal with it, so an inlined type family might work. It would be easier for people to read the code, for sure.